File: tiq-e/strings/3/main.ts
import {performance} from 'perf_hooks';
import {log} from "util";

function firstUniqChar(s: string): number {
    // for (let i = 0; i < s.length; i++) {
    //     if (s.indexOf(s[i]) === s.lastIndexOf(s[i])) {
    //         return i;
    //     }
    // }
    // return -1;
    let m = new Map()
    for (let i = 0; i < s.length; i++) {
        if (m.has(s[i])) {
            m.set(s[i], m.get(s[i])+ 1)
        } else if (!m.has(s[i])) {
            m.set(s[i], 1);
        }
    }
    for (let i = 0; i < s.length; i++) {
        if (m.get(s[i]) === 1) {
            return i
        }
    }
    return -1
};

let s = "loveleetcode"

let start = performance.now();
let result = firstUniqChar(s);
let end = performance.now();
console.log('Duration:', end - start);
console.log('Result:', result)

File: tiq-e/strings/3/main.rs
use std::collections::HashMap;
use std::time::Instant;

struct Solution;

impl Solution {
    pub fn first_uniq_char(s: String) -> i32 {
        let mut m: HashMap<char, i32> = HashMap::new();
        for l in s.chars() {
            if m.contains_key(&l) {
                m.insert(l, m.get(&l).unwrap() + 1);
            } else {
                m.insert(l, 1);
            }
        }
        for (i, l) in s.chars().enumerate() {
            if *m.get(&l).unwrap() == 1 {
                return i as i32;
            }
        }
        return -1;
    }
}

fn main() {
    let s: String = "loveleetcode".to_string();

    let start = Instant::now();
    let result = Solution::first_uniq_char(s);
    let end = start.elapsed().as_micros();
    println!("Duration: {:?}", end);
    println!("Result: {:?}", result);
}
